File: db/cost_stage_resolver.go
package database

import (
	"strings"
	"time"
)

const ContextLengthStageType = "context_length"

type TokenCostStageResolutionRequest struct {
	Model               string
	TokenType           string
	StageType           string
	RequestTime         time.Time
	InputTokensForStage int // stored request input_token_count
	TokensToBill        int // prompt/cached/output tokens to price
}

type TokenCostStageResolutionResult struct {
	Cost      float64
	Currency  string // empty when missing
	Unit      string // "1M"/"1K" (or empty if missing)
	UsedCost  *Costs
	Missing   bool
	Ambiguous bool
}

type RequestCostStageResolutionRequest struct {
	Model                 string
	RequestTime           time.Time
	InputTokenCount       int
	CachedInputTokenCount int
	OutputTokenCount      int
	StageType             string
}

type RequestCostStageResolutionResult struct {
	TotalCost float64
	Currency  string // empty when missing or currency mismatch
	Missing   bool

	InputCost  TokenCostStageResolutionResult
	CachedCost TokenCostStageResolutionResult
	OutputCost TokenCostStageResolutionResult
}

func normalizeKey(s string) string {
	return strings.TrimSpace(strings.ToLower(s))
}

func canonicalTokenTypeKey(tokenType string) string {
	t := normalizeKey(tokenType)
	switch t {
	case "input", "prompt", "input_tokens", "prompt_tokens", "inp":
		return "input"
	case "cached", "cache", "cached_input", "input_cached", "cached_input_tokens", "cached_input_token":
		return "cached"
	case "output", "completion", "output_tokens", "completion_tokens", "outp":
		return "output"
	default:
		return t
	}
}

func unitDivisor(unit string) float64 {
	if strings.TrimSpace(unit) == "1K" {
		return 1_000
	}
	// Default: treat unknown/null as "1M" to avoid exploding on partial data.
	return 1_000_000
}

func priceToCurrency(priceCents int) float64 {
	return float64(priceCents) / 100
}

func stageContains(stageMin int, stageMax *int, inputTokens int) bool {
	if inputTokens < stageMin {
		return false
	}
	if stageMax == nil {
		return true
	}
	return inputTokens <= *stageMax
}

// ResolveTokenCostStage resolves the single best cost stage row for a token type
// within the newest valid_from <= requestTime, then the stage whose range contains
// InputTokensForStage.
//
// Missing:
// - true when TokensToBill > 0 but no stage row matches.
// - false when TokensToBill <= 0 (cost is 0, stage selection is irrelevant).
// Ambiguous:
// - true when multiple stage rows match with identical stage bounds but conflicting prices/currency.
func ResolveTokenCostStage(costRows []Costs, req TokenCostStageResolutionRequest) TokenCostStageResolutionResult {
	if req.StageType == "" {
		req.StageType = ContextLengthStageType
	}
	if req.TokensToBill <= 0 {
		return TokenCostStageResolutionResult{
			Cost:      0,
			Currency:  "",
			Unit:      "",
			UsedCost:  nil,
			Missing:   false,
			Ambiguous: false,
		}
	}

	modelKey := normalizeKey(req.Model)
	tokenKey := canonicalTokenTypeKey(req.TokenType)
	stageTypeKey := normalizeKey(req.StageType)

	candidates := make([]Costs, 0, len(costRows))
	for _, c := range costRows {
		if normalizeKey(c.ModelName) != modelKey {
			continue
		}
		if canonicalTokenTypeKey(c.TokenType) != tokenKey {
			continue
		}
		if normalizeKey(c.StageType) != stageTypeKey {
			continue
		}
		if c.RequestTime.After(req.RequestTime) {
			continue
		}
		candidates = append(candidates, c)
	}

	if len(candidates) == 0 {
		return TokenCostStageResolutionResult{
			Missing:   true,
			Ambiguous: false,
		}
	}

	// 1) Select newest valid_from <= request_time
	latestValidFrom := candidates[0].RequestTime
	for _, c := range candidates[1:] {
		if c.RequestTime.After(latestValidFrom) {
			latestValidFrom = c.RequestTime
		}
	}

	candidatesAtLatest := make([]Costs, 0, len(candidates))
	for _, c := range candidates {
		if c.RequestTime.Equal(latestValidFrom) {
			candidatesAtLatest = append(candidatesAtLatest, c)
		}
	}

	// 2) Select stage whose range contains input token count
	matchingStages := make([]Costs, 0, len(candidatesAtLatest))
	for _, c := range candidatesAtLatest {
		if stageContains(c.StageMinTokens, c.StageMaxTokens, req.InputTokensForStage) {
			matchingStages = append(matchingStages, c)
		}
	}
	if len(matchingStages) == 0 {
		return TokenCostStageResolutionResult{
			Missing:   true,
			Ambiguous: false,
		}
	}

	// Deterministic tie-break:
	// - highest stage_min_tokens wins
	// - then lowest stage_max_tokens wins (treat NULL as +Inf)
	maxInt := int(^uint(0) >> 1)
	best := matchingStages[0]
	for _, c := range matchingStages[1:] {
		bestMax := maxInt
		if best.StageMaxTokens != nil {
			bestMax = *best.StageMaxTokens
		}
		cMax := maxInt
		if c.StageMaxTokens != nil {
			cMax = *c.StageMaxTokens
		}

		if c.StageMinTokens > best.StageMinTokens {
			best = c
			continue
		}
		if c.StageMinTokens == best.StageMinTokens && cMax < bestMax {
			best = c
			continue
		}
	}

	// If multiple rows have the exact same stage bounds, ensure they agree on the price.
	var (
		conflicts    = 0
		bestUnit     = strings.TrimSpace(best.UnitOfMeasure)
		bestCurrency = strings.TrimSpace(best.Currency)
	)
	for _, c := range matchingStages {
		cStageMax := c.StageMaxTokens
		sameMax := (best.StageMaxTokens == nil && cStageMax == nil) ||
			(best.StageMaxTokens != nil && cStageMax != nil && *best.StageMaxTokens == *cStageMax)
		if c.StageMinTokens == best.StageMinTokens && sameMax {
			if c.RetailPrice != best.RetailPrice ||
				strings.TrimSpace(c.UnitOfMeasure) != bestUnit ||
				strings.TrimSpace(c.Currency) != bestCurrency {
				conflicts++
			}
		}
	}
	if conflicts > 0 {
		return TokenCostStageResolutionResult{
			Missing:   true,
			Ambiguous: true,
		}
	}

	divisor := unitDivisor(best.UnitOfMeasure)
	cost := (float64(req.TokensToBill) / divisor) * priceToCurrency(best.RetailPrice)
	return TokenCostStageResolutionResult{
		Cost:      cost,
		Currency:  strings.TrimSpace(best.Currency),
		Unit:      strings.TrimSpace(best.UnitOfMeasure),
		UsedCost:  &best,
		Missing:   false,
		Ambiguous: false,
	}
}

func ResolveRequestCostStage(costRows []Costs, req RequestCostStageResolutionRequest) RequestCostStageResolutionResult {
	if req.StageType == "" {
		req.StageType = ContextLengthStageType
	}

	promptTokens := req.InputTokenCount - req.CachedInputTokenCount
	if promptTokens < 0 {
		promptTokens = 0
	}
	cachedTokens := req.CachedInputTokenCount
	if cachedTokens < 0 {
		cachedTokens = 0
	}
	outputTokens := req.OutputTokenCount
	if outputTokens < 0 {
		outputTokens = 0
	}

	inputRes := ResolveTokenCostStage(costRows, TokenCostStageResolutionRequest{
		Model:               req.Model,
		TokenType:           "input",
		StageType:           req.StageType,
		RequestTime:         req.RequestTime,
		InputTokensForStage: req.InputTokenCount,
		TokensToBill:        promptTokens,
	})
	cachedRes := ResolveTokenCostStage(costRows, TokenCostStageResolutionRequest{
		Model:               req.Model,
		TokenType:           "cached",
		StageType:           req.StageType,
		RequestTime:         req.RequestTime,
		InputTokensForStage: req.InputTokenCount,
		TokensToBill:        cachedTokens,
	})
	outputRes := ResolveTokenCostStage(costRows, TokenCostStageResolutionRequest{
		Model:               req.Model,
		TokenType:           "output",
		StageType:           req.StageType,
		RequestTime:         req.RequestTime,
		InputTokensForStage: req.InputTokenCount,
		TokensToBill:        outputTokens,
	})

	// Missing if any token type that has >0 tokens can't be resolved.
	missing := inputRes.Missing || cachedRes.Missing || outputRes.Missing
	if missing {
		return RequestCostStageResolutionResult{
			TotalCost:  0,
			Currency:   "",
			Missing:    true,
			InputCost:  inputRes,
			CachedCost: cachedRes,
			OutputCost: outputRes,
		}
	}

	// Currency selection: all resolved token parts must agree on currency.
	currencies := make([]string, 0, 3)
	if inputRes.UsedCost != nil && inputRes.Currency != "" {
		currencies = append(currencies, inputRes.Currency)
	}
	if cachedRes.UsedCost != nil && cachedRes.Currency != "" {
		currencies = append(currencies, cachedRes.Currency)
	}
	if outputRes.UsedCost != nil && outputRes.Currency != "" {
		currencies = append(currencies, outputRes.Currency)
	}

	var currency string
	if len(currencies) > 0 {
		allSame := true
		for i := 1; i < len(currencies); i++ {
			if currencies[i] != currencies[0] {
				allSame = false
				break
			}
		}
		if allSame {
			currency = currencies[0]
		}
	}

	total := inputRes.Cost + cachedRes.Cost + outputRes.Cost
	return RequestCostStageResolutionResult{
		TotalCost:  total,
		Currency:   currency,
		Missing:    false,
		InputCost:  inputRes,
		CachedCost: cachedRes,
		OutputCost: outputRes,
	}
}

File: app/src/server/costStageResolver.ts
import type { BillingUnit, RequestType } from "~/lib/costs";
import {
	type CostTokenType,
	type CostUnit,
	canonicalizeCostTokenType,
} from "~/lib/costs";

const ContextLengthStageType = "context_length" as const;

type CanonicalTokenType = CostTokenType;

export type CostStageRow = {
	model: string;
	requestType?: RequestType;
	billingUnit?: BillingUnit;
	tokenType: string;
	price: number;
	validFrom: string | Date;
	unitOfMessure: CostUnit | null;
	currency: string | null;
	stageType: string;
	stageMinTokens: number;
	stageMaxTokens: number | null;
};

export type TokenCostStageResolutionResult = {
	cost: number;
	currency: string | null; // empty when missing (handled via `missing`)
	unit: CostUnit | null;
	usedCost: CostStageRow | null;
	missing: boolean;
	ambiguous: boolean;
};

export type RequestCostStageResolutionResult = {
	totalCost: number;
	currency: string | null;
	missing: boolean;
	inputCost: TokenCostStageResolutionResult;
	cachedCost: TokenCostStageResolutionResult;
	cacheWriteCost: TokenCostStageResolutionResult;
	outputCost: TokenCostStageResolutionResult;
};

export type RerankCostResolutionResult = {
	cost: number;
	currency: string | null;
	unit: CostUnit | null;
	usedCost: CostStageRow | null;
	missing: boolean;
	ambiguous: boolean;
};

function normalizeKey(s: string) {
	return s.trim().toLowerCase();
}

function toMs(value: string | Date | null | undefined) {
	if (!value) return 0;
	return new Date(value).getTime();
}

function unitDivisor(unit: CostUnit | null) {
	if (unit === "1K") return 1_000;
	// Default: treat unknown/null as "1M" to avoid exploding on partial data.
	return 1_000_000;
}

function priceToCurrency(priceCents: number) {
	return priceCents / 100;
}

function stageContains(
	stageMin: number,
	stageMax: number | null,
	inputTokens: number,
) {
	if (inputTokens < stageMin) return false;
	if (stageMax === null) return true;
	return inputTokens <= stageMax;
}

function tokenTypeLabelWithStage(
	tokenType: CanonicalTokenType,
	stageMin: number,
	stageMax: number | null,
) {
	return stageMax === null
		? `${tokenType} (${stageMin}..∞)`
		: `${tokenType} (${stageMin}..${stageMax})`;
}

export function buildCostStageIndex(costRows: CostStageRow[]) {
	const index = new Map<string, CostStageRow[]>();

	for (const row of costRows) {
		const modelKey = normalizeKey(row.model);
		const tokenKey =
			canonicalizeCostTokenType(row.tokenType) ?? normalizeKey(row.tokenType);
		const stageTypeKey = normalizeKey(row.stageType || ContextLengthStageType);

		// tokenKey is canonicalized to input/cached/output when possible.
		const requestType = row.requestType ?? "CHAT_COMPLETION";
		const billingUnit = row.billingUnit ?? "TOKENS";
		const key = `${modelKey}|${requestType}|${billingUnit}|${tokenKey}|${stageTypeKey}`;
		const existing = index.get(key) ?? [];
		existing.push(row);
		index.set(key, existing);
	}

	return index;
}

export function resolveTokenCostStage(
	costRowsForKey: CostStageRow[],
	opts: {
		requestTime: Date;
		stageType?: string;
		inputTokensForStage: number;
		tokensToBill: number;
	},
): TokenCostStageResolutionResult {
	if (opts.tokensToBill <= 0) {
		return {
			cost: 0,
			currency: null,
			unit: null,
			usedCost: null,
			missing: false,
			ambiguous: false,
		};
	}

	if (!costRowsForKey.length) {
		return {
			cost: 0,
			currency: null,
			unit: null,
			usedCost: null,
			missing: true,
			ambiguous: false,
		};
	}

	const requestTimeMs = opts.requestTime.getTime();

	// 1) Select newest valid_from <= request_time
	let latestValidFromMs = -Infinity;
	for (const c of costRowsForKey) {
		const validFromMs = toMs(c.validFrom);
		if (validFromMs <= requestTimeMs && validFromMs > latestValidFromMs) {
			latestValidFromMs = validFromMs;
		}
	}

	if (!Number.isFinite(latestValidFromMs)) {
		return {
			cost: 0,
			currency: null,
			unit: null,
			usedCost: null,
			missing: true,
			ambiguous: false,
		};
	}

	const candidatesAtLatest = costRowsForKey.filter(
		(c) => toMs(c.validFrom) === latestValidFromMs,
	);

	// 2) Select stage whose range contains request input_token_count
	const matchingStages = candidatesAtLatest.filter((c) =>
		stageContains(c.stageMinTokens, c.stageMaxTokens, opts.inputTokensForStage),
	);

	if (!matchingStages.length) {
		return {
			cost: 0,
			currency: null,
			unit: null,
			usedCost: null,
			missing: true,
			ambiguous: false,
		};
	}

	// Deterministic tie-break:
	// - highest stage_min_tokens wins
	// - then lowest stage_max_tokens wins (treat NULL as +Inf)
	let best = matchingStages[0];
	if (!best) {
		return {
			cost: 0,
			currency: null,
			unit: null,
			usedCost: null,
			missing: true,
			ambiguous: false,
		};
	}
	let bestMax = best.stageMaxTokens ?? Number.POSITIVE_INFINITY;

	for (const c of matchingStages.slice(1)) {
		const cMax = c.stageMaxTokens ?? Number.POSITIVE_INFINITY;

		if (c.stageMinTokens > best.stageMinTokens) {
			best = c;
			bestMax = cMax;
			continue;
		}
		if (c.stageMinTokens === best.stageMinTokens && cMax < bestMax) {
			best = c;
			bestMax = cMax;
		}
	}

	// If multiple rows have the exact same stage bounds, ensure they agree on price/unit/currency.
	const bestUnit = best.unitOfMessure ?? null;
	const bestCurrency = best.currency?.trim() ?? null;

	let conflicts = 0;
	for (const c of matchingStages) {
		const sameMax =
			(best.stageMaxTokens === null && c.stageMaxTokens === null) ||
			(best.stageMaxTokens !== null &&
				c.stageMaxTokens !== null &&
				best.stageMaxTokens === c.stageMaxTokens);

		if (c.stageMinTokens === best.stageMinTokens && sameMax) {
			const cUnit = c.unitOfMessure ?? null;
			const cCurrency = c.currency?.trim() ?? null;

			if (
				c.price !== best.price ||
				cUnit !== bestUnit ||
				cCurrency !== bestCurrency
			) {
				conflicts += 1;
			}
		}
	}

	if (conflicts > 0) {
		return {
			cost: 0,
			currency: null,
			unit: null,
			usedCost: null,
			missing: true,
			ambiguous: true,
		};
	}

	const divisor = unitDivisor(best.unitOfMessure);
	const cost = (opts.tokensToBill / divisor) * priceToCurrency(best.price);

	return {
		cost,
		currency: best.currency?.trim() ?? null,
		unit: best.unitOfMessure ?? null,
		usedCost: best,
		missing: false,
		ambiguous: false,
	};
}

export function resolveRequestCostStage(
	costStageIndex: Map<string, CostStageRow[]>,
	req: {
		model: string;
		requestTime: Date;
		inputTokenCount: number;
		cachedInputTokenCount: number;
		cacheWriteTokenCount: number;
		outputTokenCount: number;
		stageType?: string;
	},
): RequestCostStageResolutionResult {
	const stageType = req.stageType?.trim() || ContextLengthStageType;

	const cachedTokens = Math.max(0, req.cachedInputTokenCount);
	const cacheWriteTokens = Math.max(0, req.cacheWriteTokenCount);
	const promptTokens = Math.max(
		0,
		req.inputTokenCount - cachedTokens - cacheWriteTokens,
	);
	const outputTokens = Math.max(0, req.outputTokenCount);

	const modelKey = normalizeKey(req.model);
	const stageTypeKey = normalizeKey(stageType);

	const resolveForTokenType = (
		tokenType: CanonicalTokenType,
		tokensToBill: number,
	) => {
		const key = `${modelKey}|CHAT_COMPLETION|TOKENS|${tokenType}|${stageTypeKey}`;
		const costRowsForKey = costStageIndex.get(key) ?? [];
		return resolveTokenCostStage(costRowsForKey, {
			requestTime: req.requestTime,
			inputTokensForStage: req.inputTokenCount,
			tokensToBill,
		});
	};

	const inputCost = resolveForTokenType("input", promptTokens);
	const cachedCost = resolveForTokenType("cached", cachedTokens);
	const cacheWriteCost = resolveForTokenType("cache_write", cacheWriteTokens);
	const outputCost = resolveForTokenType("output", outputTokens);

	const missing =
		inputCost.missing ||
		cachedCost.missing ||
		cacheWriteCost.missing ||
		outputCost.missing;
	if (missing) {
		return {
			totalCost: 0,
			currency: null,
			missing: true,
			inputCost,
			cachedCost,
			cacheWriteCost,
			outputCost,
		};
	}

	const currencies = [
		inputCost.usedCost?.currency,
		cachedCost.usedCost?.currency,
		cacheWriteCost.usedCost?.currency,
		outputCost.usedCost?.currency,
	]
		.filter((c): c is string => Boolean(c))
		.map((c) => c.trim());

	let currency: string | null = null;
	if (currencies.length > 0) {
		const first = currencies[0];
		if (first !== undefined && currencies.every((c) => c === first)) {
			currency = first;
		}
	}

	return {
		totalCost:
			inputCost.cost + cachedCost.cost + cacheWriteCost.cost + outputCost.cost,
		currency,
		missing: false,
		inputCost,
		cachedCost,
		cacheWriteCost,
		outputCost,
	};
}

export function resolveRerankCost(
	costRows: CostStageRow[],
	request: {
		model: string;
		requestTime: Date;
		searchUnits: number;
	},
): RerankCostResolutionResult {
	if (request.searchUnits <= 0) {
		return {
			cost: 0,
			currency: null,
			unit: null,
			usedCost: null,
			missing: false,
			ambiguous: false,
		};
	}

	const candidates = costRows.filter(
		(row) =>
			normalizeKey(row.model) === normalizeKey(request.model) &&
			(row.requestType ?? "CHAT_COMPLETION") === "RERANK" &&
			(row.billingUnit ?? "TOKENS") === "SEARCHES" &&
			toMs(row.validFrom) <= request.requestTime.getTime(),
	);
	if (!candidates.length) {
		return {
			cost: 0,
			currency: null,
			unit: null,
			usedCost: null,
			missing: true,
			ambiguous: false,
		};
	}

	const latest = Math.max(...candidates.map((row) => toMs(row.validFrom)));
	const matching = candidates.filter((row) => toMs(row.validFrom) === latest);
	const best = matching[0];
	if (!best) {
		return {
			cost: 0,
			currency: null,
			unit: null,
			usedCost: null,
			missing: true,
			ambiguous: false,
		};
	}
	const ambiguous = matching.some(
		(row) =>
			row.price !== best.price ||
			row.unitOfMessure !== best.unitOfMessure ||
			(row.currency ?? null) !== (best.currency ?? null),
	);
	if (ambiguous) {
		return {
			cost: 0,
			currency: null,
			unit: null,
			usedCost: null,
			missing: true,
			ambiguous: true,
		};
	}

	return {
		cost:
			(request.searchUnits / unitDivisor(best.unitOfMessure)) *
			priceToCurrency(best.price),
		currency: best.currency?.trim() ?? null,
		unit: best.unitOfMessure ?? null,
		usedCost: best,
		missing: false,
		ambiguous: false,
	};
}

export { tokenTypeLabelWithStage };
